### DISCUSSION TITLE
minimum requirements to mock HttpContext.GetRequestedApiVersion()

### DISCUSSION THREAD

**Discussion by steveoh** [dotnet/aspnet-api-versioning#981] [Q&A] [answered]
I'm wanting to test some functionality and setting up my test incorrectly so I'm asking for a little guidance.

I have a little helper to create an `HttpContext`

```c#
public static HttpContext CreateVersionedHttpContext(int version) {
    var defaultHttpContext = new DefaultHttpContext();

    defaultHttpContext.Features.Set(new ApiVersioningFeature(defaultHttpContext) {
        RequestedApiVersion = new ApiVersion(version, 0)
    });

    return defaultHttpContext;
}
```

In my test I return that value from an injectable `IHttpContextAccessor`

```c#
[Fact]
public void Should_return_v2_for_api_v2() {
    var httpContext = HttpContextHelpers.CreateVersionedHttpContext(2);

    var httpContextAccessor = new Mock<IHttpContextAccessor>();
    httpContextAccessor.SetupGet(x => x.HttpContext).Returns(httpContext);

   // ...
```

In the class I'm testing I inject the `IHttpContextAccessor` and try to use the `GetRequestedApiVersion` extension method but it returns null.

```c#
   var version = _httpContext.GetRequestedApiVersion();
```

If I look at the features of the `_httpContext` I can see the version values are there. But I must be missing something. I'm going to dig through the source code to see what the extension method is doing to see if that will offer any clues while you get this notification :) 

![image](https://user-images.githubusercontent.com/325813/228402166-eea7a3ca-3a4c-49b2-ae88-39840ac121d7.png)

[These tests](https://github.com/dotnet/aspnet-api-versioning/blob/main/src/AspNetCore/WebApi/test/Asp.Versioning.Http.Tests/Http/HttpContextExtensionsTest.cs) are helpful but they seem to setup a lot of things and are using the query string only. Are there tests for the other ways to set the version?

This seems to work.

```c#
var version = new ApiVersion(2, 0);
var feature = new Mock<IApiVersioningFeature>();
var featureCollection = new Mock<IFeatureCollection>();
var httpContext = new Mock<HttpContext>();

feature.SetupProperty(f => f.RequestedApiVersion, version);
featureCollection.Setup(fc => fc.Get<IApiVersioningFeature>()).Returns(feature.Object);
httpContext.SetupGet(hc => hc.Features).Returns(featureCollection.Object);

var httpContextAccessor = new Mock<IHttpContextAccessor>();
httpContextAccessor.SetupGet(x => x.HttpContext).Returns(httpContext.Object);
```

Is there a suggested way for when using the `DefaultHttpContext` and not mocking the entire thing?

**Comment by commonsensesoftware** [comment, marked as answer]
I'm glad you found the tests. That _is_ the documentation. 😉 

Setting up and parsing from the `HttpContext` is irrelevant and unnecessary. The specific tests you mention exercise that path for coverage, but it is unnecessary. I wouldn't recommend you _re-test_ that behavior. If all you want to do is test the behavior for a specific `ApiVersion`, you can explicitly set it in the feature, which it seems you've figured out.

You've missed just one very small, but important piece. Wait for it.... Warning: You are going to kick yourself... 2 ... 3:

```c#
public static HttpContext CreateVersionedHttpContext(int version) {
    var context = new DefaultHttpContext();
    IApiVersionVersioningFeature feature = new ApiVersioningFeature(context)
    {
        RequestedApiVersion = new ApiVersion(version, 0),
    };
    context.Features.Set(feature);
    return context;
}
```

Hopefully, you've spotted the issue, but if not, the problem is that you inlined `new ApiVersioningFeature(...)` in `Features.Set` which will use the type `ApiVersioningFeature`  as the key. That is wrong. The key is type `IApiVersioningFeature` as shown in the test. There are several ways to address it:

1. Use a correctly-typed variable `IApiVersioningFeature feature` as show above
2. Explicitly set the type as in the tests - `context.Features.Set<IApiVersioningFeature>(new ApiVersioningFeature(context))`
3. Use `context.ApiVersioningFeature()` instead

`#3` is shorthand to avoid having to repeatedly check if the feature is present and add it. If no other setup is provided, it will test and create the default feature instance on the current `IFeatureCollection` as necessary. This will ensure it's always setup correctly. For example:

```c#
public static HttpContext CreateVersionedHttpContext(int version) {
    var context = new DefaultHttpContext();
    context.ApiVersioningFeature().RequestedApiVersion = new(version, 0);
    return context;
}
```

I hope that helps.

**Reply by steveoh** [reply]
savage!
